### PR TITLE
endpoint: Fix realized state corruption when initial policy regeneration fails

### PR DESCRIPTION
**Pull request by tgraf**
When an endpoint regeneration calculates policy for the first time, the policy
map is created. Up to now, the regeneration function queued a deferred function
that, on error, deleted the policy map again. If the endpoint generation failed
*after* the policy map content was already successfully synchronization, the
removal of the policy map result in a flush of the policy map content without
resetting e.realizedMapState which in turn would make e.syncPolicyMap() believe
that the desired state was already fully realized when the BPF map itself was
flushed. This would result in the BPF map being out of sync with what is
believed to be the realized state.

There is no need to defer the deletion of the map if the regeneration fails,
the policy map can remain in the endpoint. It will get reused in the next
regeneration attempt and will eventually get deleted when the endpoint is
removed together with all other BPF maps tied to the endpoint.

Similarly, if the policy map is removed for some reason, the regeneration does
re-create the map but did not reset the e.realizedMapState and thus would never
write to the map. Resolve this by always resetting the realizedMapState when
the policy map is created.

Signed-off-by: Thomas Graf <thomas@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5483)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/endpoint/bpf.go [hunk 1]
```diff
@@ -424,26 +424,8 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
...
-				os.RemoveAll(e.PolicyMapPathLocked())
-				e.PolicyMap = nil
-				e.Unlock()
-			}
-		}
-	}()
-
 	if e.PolicyMap == nil {
-		e.PolicyMap, createdPolicyMap, err = policymap.OpenMap(e.PolicyMapPathLocked())
+		e.PolicyMap, _, err = policymap.OpenMap(e.PolicyMapPathLocked())
```
Comment (rlenglet) [inline pkg/endpoint/bpf.go:428]: This was the last call to `policymap.OpenMap` that made use of the returned "new map" boolean.
That can be removed from `policymap.OpenMap`.

**Review by rlenglet — APPROVED** [review]

File: pkg/endpoint/bpf.go (hunk 1 above)
Comment (rlenglet) [reply pkg/endpoint/bpf.go:428]: More generally, it seems like that the "new map" boolean returned by `bpf.OpenOrCreateMap` is not needed anymore.

Comment (tgraf) [reply pkg/endpoint/bpf.go:428]: Agreed. I kept the fix minimal as we backport this to all branches.

**Review by ianvernon — APPROVED** [review]

**Pull request merged**